cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/query/FlowQueries.kt

```diff
@@ -0,0 +1,654 @@
+ * (unlimited depth if `null`). [maxSteps] defines the total number of steps we will follow in the
+ * graph (unlimited depth if `null`).
+ */
+class Interprocedural(val maxCallDepth: Int? = null, maxSteps: Int? = null) :
+    AnalysisScope(maxSteps) {
+    override fun followEdge(
+        currentNode: Node,
+        edge: Edge<Node>,
+        ctx: Context,
+        analysisDirection: AnalysisDirection?,
+    ): Boolean {
+        // Follow the edge if we're still in the maxSteps range and (if maxCallDepth is null or the
+        // call stack is not deeper yet)
+        return (this.maxSteps == null || ctx.steps < maxSteps) &&
+            (maxCallDepth == null || ctx.callStack.depth < maxCallDepth)
+    }
+}
+
+/** Determines in which direction we follow the edges. */
+sealed class AnalysisDirection : StepSelector {
+    override fun followEdge(
+        currentNode: Node,
+        edge: Edge<Node>,
+        ctx: Context,
+        analysisDirection: AnalysisDirection?,
+    ): Boolean {
+        return true
+    }
+
+    abstract fun unwrapNextStepFromEdge(edge: Edge<Node>): Node
+}
+
+/** Follow the order of the EOG */
+class Forward() : AnalysisDirection() {
+    override fun unwrapNextStepFromEdge(edge: Edge<Node>): Node {
+        return edge.end
+    }
+}
+
+/** Against the order of the EOG */
+class Backward() : AnalysisDirection() {
+    override fun unwrapNextStepFromEdge(edge: Edge<Node>): Node {
+        return edge.start
+    }
+}
+
+/** In and against the order of the EOG */
+class Bidirectional() : AnalysisDirection() {
+    override fun unwrapNextStepFromEdge(edge: Edge<Node>): Node {
+        TODO("Not yet implemented")
+    }
+}
+
+/** Determines if the predicate must or may hold */
+enum class AnalysisType {
+    /**
+     * The predicate must hold, i.e., all paths fulfill the property/requirement. No path violates
+     * the property/requirement.
+     */
+    MUST,
+    /**
+     * The predicate may hold, i.e., there is at least one path which fulfills the
+     * property/requirement.
+     */
+    MAY,
+}
+
+/** Configures the sensitivity of the analysis. */
+sealed class AnalysisSensitivity : StepSelector {
+    operator fun plus(other: AnalysisSensitivity) = arrayOf(this, other)
+
+    operator fun List<AnalysisSensitivity>.plus(
+        other: AnalysisSensitivity
+    ): Array<AnalysisSensitivity> {
+        return arrayOf(*this.toTypedArray(), other)
+    }
+
+    operator fun plus(other: Array<AnalysisSensitivity>): Array<AnalysisSensitivity> {
+        return arrayOf(*other, this)
+    }
+}
+
+/** Consider the calling context when following paths (e.g. based on a call stack). */
+class ContextSensitive() : AnalysisSensitivity() {
+    override fun followEdge(
+        currentNode: Node,
+        edge: Edge<Node>,
+        ctx: Context,
+        analysisDirection: AnalysisDirection?,
+    ): Boolean {
+        return if (edge is ContextSensitiveDataflow && analysisDirection is Forward) {
```

Comment:
Currently, this only works for `Dataflow` edges but we also need something for the `EvaluationOrder`. However, it will probably be somewhat different...

I also dislike that the code for `Forward` and `Backward` looks similar with the only difference being when to use `CallingContextIn` and `CallingContextOut`. Good ideas on how to improve this are welcome